packages/calling-stateful-client/src/TypeGuards.ts

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Call, CallAgent } from '@azure/communication-calling';
+import { CallAgentCommon, CallCommon, TeamsCall, TeamsCallAgent } from './BetaToStableTypes';
+
+/**
+ * @private
+ */
+export const isACSCall = (call: CallCommon): call is Call => {
+  /* @conditional-compile-remove(teams-identity-support) */
+  return call.kind === 'Call';
+  return true;
+};
+
+/**
+ * @private
+ */
+export const isACSCallAgent = (callAgent: CallAgentCommon): callAgent is CallAgent => {
+  /* @conditional-compile-remove(teams-identity-support) */
+  return callAgent.kind === 'CallAgent';
+  return true;
+};
+
+/**
+ * @private
+ */
+export const isTeamsCall = (call: CallCommon): call is TeamsCall => {
+  /* @conditional-compile-remove(teams-identity-support) */
+  return call.kind === 'TeamsCall';
+  return false;
+};
+
+/**
+ * @private
+ */
+export const isTeamsCallAgent = (callAgent: CallAgentCommon): callAgent is TeamsCallAgent => {
+  /* @conditional-compile-remove(teams-identity-support) */
+  return callAgent.kind === 'TeamsCallAgent';
```

Comment:
My understanding is this technically isn't true. `callAgent` here is of type `CallAgentCommon`, but CallAgentCommon is not a TeamsCallAgent:
![image](https://user-images.githubusercontent.com/2684369/199279754-8927eb7c-364c-45c9-965c-422ad7c66909.png)
Although _we_ always are using it as either a (ACS)CallAgent or TeamsCallAgent this function could take any call agent common and it will only check the `kind` property then incorrectly typecast the entire object.

(This is a continuation of this comment: https://github.com/Azure/communication-ui-library/pull/2396/files#r1010585946 where I'm interested to know why we use `CallAgentCommon` and not `CallAgent | TeamsCallAgent` instead)

Reply:
This is a good point which I was thinking of for a quite while.

I was using CallAgent | TeamsCallAgent as parameter at the very beginning, but found it a pain to implement it in an elegant way. This was actually modified after I had first iteration review with @prprabhu-ms. There are a few reason we use CallAgentCommon here:

1. Using CallAgent | TeamsCallAgent as parameters requires us to pass it from parent type as CallAgent | TeamsCallAgent, which means there must be a type cast somewhere from the caller (cause it should be callAgentCommon at very top level), which means this type guard loses 50% of its functionality. (and type cast will highly depend on .kind property still)

2. This is a fairly common trick to do both type check both in js and ts. Kind property exists in callCommon, and the only reason it exists there is for type cast(or we say runtime type check for js), if the kind value indicates a wrong type, I would say it is a serious bug coming from our calling sdk

In short, any assumption to check one or two properties then do type cast is not safe in the end, unless you check all the properties there and check all their type. So this is just another "better typescript version" of js runtime check

Reply:
Can you point to where the typecast from `CallAgent | TeamsCallAgent` to `CallAgentCommon` would be needed?
`CallAgent` and `TeamsCallAgent` should be a superset of `CallAgentCommon` (they should extend call agent common) so anywhere requiring a CallAgentCommon could take a CallAgent or TeamsCallAgent without issue no?

Reply:
Sry, my description above isn't clear enough.

It's not casting CallAgent | TeamsAgent to AgentCommon
It's casting AgentCommon to CallAgent | TeamsAgent

Because at some point of ancestor of functions who calls isACSAgent, the type is CallAgentCommon (because everything start with common code then we make it specify later using if condition)

If this type guard uses CallAgent | TeamsAgent, that means we will have to cast CallAgentCommon to CallAgent | TeamsAgent before we call this typeGuard(which is actually what this type guard for), otherwise it wouldn't satisfy type check

Reply:
One way to make this work is never using CallAgentCommon in ancestor but that makes typing much complicated than this one(you can check my first several commit, which I used CallAgent | TeamsAgent for these functions)

Reply:
Have to step out but will be back in 30mins, Looking at the old commits:
![image](https://user-images.githubusercontent.com/2684369/199823995-8ef9f1cc-ebcf-4e43-afcd-b4b108126ec1.png)

Why are we type casting in the Common, and not having the CallAgent or the TeamsCallAgent house their specific implementations?

Reply:
That's what we are doing for current implementation, you can see the only difference here is those if-elses, so we forward that child class

There will be lots of CallCommon/CallAgentCommon usage in handler and Adapter layer which requires some minor run time checks, the need of converting CallAgentCommon to CallAgent or TeamsAgent is not avoidable, 

Back to typeguard itself, that's the intention of why sdk put kind property in their common object not only in TeamsAgent | CallAgent, so this is just a way to leverage that pattern in typescript (I do think sdk should provide us the type guard function)